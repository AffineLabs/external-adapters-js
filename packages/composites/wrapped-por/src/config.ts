import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'WRAPPED_POR'
export const makeConfig = (prefix?: string): Config => Requester.getDefaultConfig(prefix)
