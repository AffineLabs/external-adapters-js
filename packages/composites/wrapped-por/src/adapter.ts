// import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { AdapterRequest, AdapterResponse, Execute, AdapterContext } from '@chainlink/types'
import * as PoR from '@chainlink/proof-of-reserves-adapter'
import * as wrapped from '@chainlink/wrapped-adapter'
import { makeConfig } from './config'

export const execute = async (
  input: AdapterRequest,
  context: AdapterContext,
): Promise<AdapterResponse> => {
  const validator = new Validator(input)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const symbol = validator.validated.protocol || 'ETH'
  const protocol = validator.validated.protocol || 'list'
  const indexer = validator.validated.indexer || 'eth_balance'

  const _executeWrapped = wrapped.makeExecute()
  const addresses = await _executeWrapped(
    {
      id: jobRunID,
      data: {
        symbol: symbol,
      },
    },
    context,
  )

  if (addresses.statusCode !== 200) {
    return addresses
  }

  const _executePoR = PoR.makeExecute()
  const sum = await _executePoR(
    {
      id: jobRunID,
      data: {
        protocol: protocol,
        indexer: indexer,
        addresses: addresses.result,
      },
    },
    context,
  )
  return Requester.success(jobRunID, sum)
}

export const makeExecute = (): Execute => {
  return async (request, context) => execute(request, context || makeConfig())
}
