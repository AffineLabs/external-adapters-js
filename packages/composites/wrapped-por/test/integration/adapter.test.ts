import { AdapterRequest } from '@chainlink/types'
import request from 'supertest'
import * as process from 'process'
import { server as startServer } from '../../src'
import { mockETHBalanceResponseSuccess } from './fixtures'
import * as nock from 'nock'
import * as http from 'http'

beforeAll(() => {
  process.env.CACHE_ENABLED = 'false'
  process.env.ETH_BALANCE_ADAPTER_URL =
    process.env.ETH_BALANCE_ADAPTER_URL || 'http://localhost:8081'
  if (process.env.RECORD) {
    nock.recorder.rec()
  }
})

afterAll(() => {
  if (process.env.RECORD) {
    nock.recorder.play()
  }

  nock.restore()
  nock.cleanAll()
  nock.enableNetConnect()
})

describe('execute', () => {
  const id = '1'
  let server: http.Server
  const req = request('localhost:8080')
  beforeAll(async () => {
    server = await startServer()
    process.env.CACHE_ENABLED = 'false'
  })
  afterAll((done) => {
    server.close(done)
  })

  describe('get symbol address sum balances', () => {
    const data: AdapterRequest = {
      id,
      data: {
        result: [{ symbol: 'ETH' }],
      },
    }

    it('should return success', async () => {
      mockETHBalanceResponseSuccess()

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)

      expect(response.body).toMatchSnapshot()
    })
  })
})
