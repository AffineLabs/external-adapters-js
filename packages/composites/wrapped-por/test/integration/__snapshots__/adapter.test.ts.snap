// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute get symbol address sum balances should return success 1`] = `
Object {
  "data": Object {
    "result": "3858950875039521157800",
  },
  "jobRunID": "1",
  "result": "3858950875039521157800",
  "statusCode": 200,
}
`;
